torch/csrc/jit/import.cpp

```diff
@@ -134,7 +137,18 @@ at::Tensor ScriptModuleDeserializer::loadTensor(const torch::TensorDef& tensor_p
   }
   auto t = at::CPU(type)._th_tensor(
       storage_it->second, tensor_proto.offset(), dims, strides);
-  return autograd::make_variable(t, tensor_proto.requires_grad());
+
+  t = autograd::make_variable(t, tensor_proto.requires_grad());
+
+  // restore to the appropriate device
+  AT_ASSERT(tensor_proto.has_device() && !tensor_proto.device().empty());
+  const std::string mapped_device_str = deviceLookup_(t, tensor_proto.device());
+  at::Device mapped_device(mapped_device_str);
+  if (mapped_device != t.device()) {
+    t = t.to(mapped_device, t.scalar_type());
+    storage_it->second = t.storage();
```

Comment:
so the device is semantically associated with storage, not tensor. You probably want to move it above where you create  the storage and here - enforce that  device is consistent between tensors with shared storages